strawberryfields/decompositions.py

```diff
@@ -476,6 +476,190 @@ def rectangular_symmetric(V, tol=1e-11):
     return new_tlist, new_diags, None
 
 
+def BS_mat(theta):
+    r"""Beam splitter matrix as implemented in hardware
+    """
+    mat = np.array([[np.cos(theta), 1j*np.sin(theta)], [1j*np.sin(theta), np.cos(theta)]])
+    return mat
+
+def R(phi):
+    r"""Phase as implemented in hardware
+    """
+    mat = np.array([[np.exp(1j*phi), 0], [0, 1]])
+    return mat
+
+def MZ(m, n, phi_int, phi_ext, nmax):
+    r"""One unit as implemented on the chip. 
+    An external phase followed by a beamsplitter followed by 
+    the internal phase followed by another beamsplitter
+    """
+    theta = np.pi/4
+    BS1 = BS_mat(theta)
+    BS2 = BS_mat(theta)
+    T = BS1 @ R(phi_int) @ BS2 @ R(phi_ext)
+    
+    mat = np.identity(nmax, dtype=np.complex128)
+    mat[m, m] = T[0][0]
+    mat[m, n] = T[0][1]
+    mat[n, m] = T[1][0]
+    mat[n, n] = T[1][1]
+    return mat
+
+
+def MZi(m, n, phi_int, phi_ext, nmax):
+    r"""The inverse MZ matrix."""
+    return np.transpose(np.conj(MZ(m, n, phi_int, phi_ext, nmax)))
+
+def nullMZi(m, n, U):
+    r"""Nullifies element m,n of U using MZi"""
+    (nmax, mmax) = U.shape
+
+    if nmax != mmax:
+        raise ValueError("U must be a square matrix")
+
+    if U[m, n] == 0 :
+        phi_i = np.pi
+        phi_e = 0
+    elif U[m, n+1] == 0:
+        phi_i = 0
+        phi_e = 0
+    else:
+        r = - U[m, n+1] / U[m, n]
+        phi_i = 2*np.arctan(np.abs(r))
+        phi_e = - np.angle(r)
+
+    return [n, n+1, phi_i, phi_e, nmax]
+
+
+def nullMZ(n, m, U):
+    r"""Nullifies element n,m of U using MZ"""
+    (nmax, mmax) = U.shape
+
+    if nmax != mmax:
+        raise ValueError("U must be a square matrix")
+
+    if U[n, m] == 0 :
+        phi_i = np.pi
+        phi_e = 0
+    elif U[n-1, m] == 0:
+        phi_i = 0
+        phi_e = 0
+    else:
+        r = U[n-1, m]/ U[n, m] 
+        phi_i = 2*np.arctan(np.abs(r))
+        phi_e = - np.angle(r)
+
+    return [n-1, n, phi_i, phi_e, nmax]
+
+def rectangular_MZ(V, tol=1e-11):
```

Comment:
Should this function be named `rectangular_symmetric`, and fully replace the existing `rectangular_symmetric` function? As, as far as I can tell, the functionality here is exactly the same as the existing `rectangular_symmetric`, but more efficient.

By renaming this function to `rectangular_symmetric` (and deleting the old one), the advantage is that the existing rectangular symmetric tests will also run on this function.